src/AstTransforms.cpp

```diff
@@ -220,37 +221,66 @@ bool MaterializeAggregationQueriesTransformer::materializeAggregationQueries(
             // check whether a materialization is required
             if (!needsMaterializedRelation(agg)) {
                 return;
             }
-
             changed = true;
 
-            // for more body literals: create a new relation
-            std::set<std::string> vars;
-            visitDepthFirst(agg, [&](const AstVariable& var) { vars.insert(var.getName()); });
-
             // -- create a new clause --
 
             auto relName = "__agg_rel_" + toString(counter++);
             while (program.getRelation(relName) != nullptr) {
                 relName = "__agg_rel_" + toString(counter++);
             }
+            // create the new clause for the materialised thing
+            auto* aggClause = new AstClause();
+            // create the body of the new thing
+            for (const auto& cur : agg.getBodyLiterals()) {
+                aggClause->addToBody(std::unique_ptr<AstLiteral>(cur->clone()));
+            }
+            // find stuff for which we need a grounding
+            for (const auto& argPair : getGroundedTerms(*aggClause)) {
+              const AstVariable* variable = dynamic_cast<const AstVariable*>(argPair.first);
+              // if it's not even a variable type or the term is grounded
+              // then skip it
+              if (variable == nullptr || argPair.second)
+                continue;
+
+              for (const auto& lit : clause.getBodyLiterals()) {
+                const AstAtom* atom = dynamic_cast<const AstAtom*>(lit);
+                if (!atom) continue; // ignore these because they can't ground the variable
+                for (const auto& arg : atom->getArguments()) {
+                    const AstVariable* atomVariable = dynamic_cast<const AstVariable*>(arg);
+                    // if this atom contains the variable I need to ground, add it
+                    if (atomVariable && variable->getName() == atomVariable->getName()) {
+                      // expand the body with this one so that it will ground this variable
+                      aggClause->addToBody(std::unique_ptr<AstLiteral>(atom->clone()));
+                      break;
+                    }
+                  }
+              }
+            }
 
+            // now all that's left to do is set the head with the
+            // appropriate parameters. (All of those that occur in the body
+            // of the aggregate)
             auto* head = new AstAtom();
             head->setName(relName);
             std::vector<bool> symbolArguments;
-            for (const auto& cur : vars) {
-                head->addArgument(std::make_unique<AstVariable>(cur));
+
+            // i just need to find a set of the variables so i don't add things twice
+            std::set<std::string> variables;
+            visitDepthFirst(*aggClause, [&](const AstVariable& var) { variables.insert(var.getName()); });
+
+            // then just put all of those into the head. they are all completely
```

Comment:
Either remove this comment, or indicate why they are all necessary. Either is good.